vim-patch:8.2.3917: the eval.txt help file is way too big

dstein64 opened the pull request:
This ports vim/vim@1cae5a0 to Neovim.

The patch was not applied as-is, having the following changes:

- The list of functions and their details for `builtin.txt` were copied from the Neovim's `eval.txt`, not Vim's.
- The header for `builtin.txt` uses Neovim's help formatting (e.g., "Nvim" on first line, no "Last change").
- There is no "Feature list" section to port (consequently, the `string-match` section number was decremented).
- ~~The "Not all functions are here" text was not added under `builtin-function-details` as it was for Vim, since that text presumably refers to Vim help files that are not present in Neovim (`popup.txt`, `term.txt`, etc.). While Neovim does have `api.txt`, the functions in that help file are not documented in `builtin-function-list`.~~
- The changes to `Makefile`, `remote.txt`, and `version.c` were not relevant for Neovim.
- Neovim-specific references to `|functions|` were changed to `|builtin-functions|`.

#### vim-patch:8.2.3917: the eval.txt help file is way too big

Problem:    The eval.txt help file is way too big.
Solution:   Move the builtin function details to a separate file.
https://github.com/vim/vim/commit/1cae5a0a034d0545360387407a7a409310f1efe2

#### vim-patch:partial:04fb916

Update runtime files
https://github.com/vim/vim/commit/04fb916684829f6aa12f33f14d0d0023b458f200

dstein64 committed:
vim-patch:8.2.3917: the eval.txt help file is way too big

Problem:    The eval.txt help file is way too big.
Solution:   Move the builtin function details to a separate file.
https://github.com/vim/vim/commit/1cae5a0a034d0545360387407a7a409310f1efe2

Note: Neovim-specific references to |functions| were changed to
|builtin-functions|. This included updates to:
  1. test/functional/vimscript/functions_spec.lua
  2. test/functional/vimscript/eval_spec.lua
  3. runtime/doc/lua.txt

clason commented:
Probably should hold off until https://github.com/neovim/neovim/pull/17007 is merged (which should be today or tomorrow).

clason commented:
And now it's merged, so there's conflicts.

dstein64 commented:
Thanks for the update @clason! I've rebased, moving the relevant update to `builtin.txt`.

```diff
diff --git a/runtime/doc/builtin.txt b/runtime/doc/builtin.txt
index eb12e6808..2be0cd39c 100644
--- a/runtime/doc/builtin.txt
+++ b/runtime/doc/builtin.txt
@@ -2318,8 +2318,10 @@ funcref({name} [, {arglist}] [, {dict}])
                function {name} is redefined later.
 
                Unlike |function()|, {name} must be an existing user function.
-               Also for autoloaded functions. {name} cannot be a builtin
-               function.
+               It only works for an autoloaded function if it has already
+               been loaded (to avoid mistakenly loading the autoload script
+               when only intending to use the function name, use |function()|
+               instead). {name} cannot be a builtin function.
 
                Can also be used as a |method|: >
                        GetFuncname()->funcref([arg])
@@ -4636,7 +4638,7 @@ match({expr}, {pat} [, {start} [, {count}]])                      *match()*
                        GetText()->match('word')
                        GetList()->match('word')
 <
-                               *matchadd()* *E798* *E799* *E801* *E957*
+                       *matchadd()* *E798* *E799* *E801* *E957*
 matchadd({group}, {pattern}[, {priority}[, {id} [, {dict}]]])
                Defines a pattern to be highlighted in the current window (a
                "match").  It will be highlighted with {group}.  Returns an
```

seandewar reviewed (commented):
looks pretty good to me i think. only missing a few small things:

- https://github.com/vim/vim/commit/1cae5a0a034d0545360387407a7a409310f1efe2#diff-e5cba4644a37135c31c9f067b5d0c5de662134c7d3a6bf5182417eec0844661fR139
- https://github.com/vim/vim/commit/1cae5a0a034d0545360387407a7a409310f1efe2#diff-ef8946d89919ad628b9ac65c49d3067e4e017b0980cac70eb25d832a59ef677dR720-R721 (this is applicable to `test_` functions, for example)
- https://github.com/vim/vim/commit/1cae5a0a034d0545360387407a7a409310f1efe2#diff-ef8946d89919ad628b9ac65c49d3067e4e017b0980cac70eb25d832a59ef677dR10106 (the patch very sneakily added `This is common between several functions.`; luckily the other added stuff was vim9 related 😆)

dstein64 commented:
Thanks @seandewar. I've made the updates and force-pushed (4db4416af7f29103c1348921be442c87fe3bb0d2). I've crossed out the text on "Not all functions are here" in the description [above](https://github.com/neovim/neovim/pull/16975#issue-1096529123).

Here's the diff:

```diff
diff --git a/runtime/doc/builtin.txt b/runtime/doc/builtin.txt
index 2be0cd39c..9dd0a95b6 100644
--- a/runtime/doc/builtin.txt
+++ b/runtime/doc/builtin.txt
@@ -546,6 +546,9 @@ xor({expr}, {expr})         Number  bitwise XOR
 ==============================================================================
 2. Details                                     *builtin-function-details*
 
+Not all functions are here, some have been moved to a help file covering the
+specific functionality.
+
 abs({expr})                                                    *abs()*
                Return the absolute value of {expr}.  When {expr} evaluates to
                a |Float| abs() returns a |Float|.  When {expr} can be
@@ -8642,12 +8645,13 @@ xor({expr}, {expr})                                     *xor()*
 ==============================================================================
 3. Matching a pattern in a String                      *string-match*
 
-A regexp pattern as explained at |pattern| is normally used to find a match in
-the buffer lines.  When a pattern is used to find a match in a String, almost
-everything works in the same way.  The difference is that a String is handled
-like it is one line.  When it contains a "\n" character, this is not seen as a
-line break for the pattern.  It can be matched with a "\n" in the pattern, or
-with ".".  Example: >
+This is common between several functions. A regexp pattern as explained at
+|pattern| is normally used to find a match in the buffer lines.  When a
+pattern is used to find a match in a String, almost everything works in the
+same way.  The difference is that a String is handled like it is one line.
+When it contains a "\n" character, this is not seen as a line break for the
+pattern.  It can be matched with a "\n" in the pattern, or with ".".  Example:
+>
        :let a = "aaaa\nxxxx"
        :echo matchstr(a, "..\n..")
        aa
diff --git a/runtime/doc/help.txt b/runtime/doc/help.txt
index 173d3c0cd..b97c9a2e3 100644
--- a/runtime/doc/help.txt
+++ b/runtime/doc/help.txt
@@ -129,6 +129,7 @@ Advanced editing ~
 |diff.txt|     working with two to eight versions of the same file
 |autocmd.txt|  automatically executing commands on an event
 |eval.txt|     expression evaluation, conditional commands
+|builtin.txt|  builtin functions
 |fold.txt|     hide (fold) ranges of lines
 |lua.txt|      Lua API
 |api.txt|      Nvim API via RPC, Lua and VimL
```

dstein64 committed:
vim-patch:partial:04fb916

Update runtime files
https://github.com/vim/vim/commit/04fb916684829f6aa12f33f14d0d0023b458f200

dstein64 commented:
> "also there were a few applicable changes made in future runtime updates that may be of interest:"

Done (30e1b9851df18305a1f677833ff5924eb5aabcc5).

(PR description updated accordingly)

seandewar reviewed (approved):
LGTM cc @clason

dstein64 commented:
> "Thanks! Can you also add these changes:"

👍 Sure. The partial updates from https://github.com/vim/vim/commit/04fb916684829f6aa12f33f14d0d0023b458f200 were made in a separate commit (same update mentioned [here](https://github.com/neovim/neovim/pull/16975#issuecomment-1011424579)). The update to change `|functions|` to `|builtin-functions|` was incorporated into the original commit, with a note added to the commit message and a corresponding update to the PR description above.

dstein64 commented on src/nvim/testdir/test_regexp_utf8.vim:
```diff
@@ -590,4 +590,12 @@ func Test_match_char_class_upper()
   bwipe!
 endfunc
 
+func Test_match_invalid_byte()
+  call writefile(0z630a.765d30aa0a.2e0a.790a.4030, 'Xinvalid')
+  new
+  source Xinvalid
+  bwipe!
+  call delete('Xinvalid')
+endfunc
+
```
I'll get rid of this. This was a added by mistake (not sure how).

seandewar replied:
maybe a messed up rebase? 🤷

dstein64 replied:
I think so but I'm not sure.

I dropped the commit that was mistakely added.

dstein64 commented:
The recent merge of PR #17028 resulted in a conflict with this PR. I've rebased and force pushed. This PR now includes documentation for `win_move_separator` and `win_move_statusline`.

dstein64 commented:
The recent merge of PR #17222 resulted in a conflict with this PR. I've rebased and force pushed. This PR now removes `python2` from the `has()` documentation in `builtin.txt`.

bfredl commented:
thanks for keeping it up to date to our conflicting changes @dstein64  :)
